Index: odhcpd-2020-05-03-49e4949c/src/odhcpd.c
===================================================================
--- odhcpd-2020-05-03-49e4949c.orig/src/odhcpd.c
+++ odhcpd-2020-05-03-49e4949c/src/odhcpd.c
@@ -218,7 +218,7 @@ ssize_t odhcpd_send(int socket, struct s
 }
 
 
-static int odhcpd_get_linklocal_interface_address(int ifindex, struct in6_addr *lladdr)
+int odhcpd_get_linklocal_interface_address(int ifindex, struct in6_addr *lladdr)
 {
 	int ret = -1;
 	struct sockaddr_in6 addr;
Index: odhcpd-2020-05-03-49e4949c/src/odhcpd.h
===================================================================
--- odhcpd-2020-05-03-49e4949c.orig/src/odhcpd.h
+++ odhcpd-2020-05-03-49e4949c/src/odhcpd.h
@@ -343,6 +343,7 @@ ssize_t odhcpd_send(int socket, struct s
 		const struct interface *iface);
 int odhcpd_get_interface_dns_addr(const struct interface *iface,
 		struct in6_addr *addr);
+int odhcpd_get_linklocal_interface_address(int ifindex, struct in6_addr *lladdr);
 int odhcpd_get_interface_config(const char *ifname, const char *what);
 int odhcpd_get_mac(const struct interface *iface, uint8_t mac[6]);
 struct interface* odhcpd_get_interface_by_index(int ifindex);
Index: odhcpd-2020-05-03-49e4949c/src/router.c
===================================================================
--- odhcpd-2020-05-03-49e4949c.orig/src/router.c
+++ odhcpd-2020-05-03-49e4949c/src/router.c
@@ -630,7 +630,10 @@ static int send_router_advert(struct int
 		if (iface->dns_cnt > 0) {
 			dns_addr = iface->dns;
 			dns_cnt = iface->dns_cnt;
-		} else if (!odhcpd_get_interface_dns_addr(iface, &dns_pref)) {
+		//} else if (!odhcpd_get_interface_dns_addr(iface, &dns_pref)) {
+		/*  The address of IPv6 Recursive DNS Server is the router’s link-local unicast
+			IPv6 address on the LAN interface, requirement from netgear spec */
+		} else if (!odhcpd_get_linklocal_interface_address(iface->ifindex, &dns_pref)) {
 			dns_addr = &dns_pref;
 			dns_cnt = 1;
 		}
