--- a/pppd/plugins/pptp/pptp.c	2021-07-15 16:44:16.098971500 +0800
+++ b/pppd/plugins/pptp/pptp.c	2021-07-21 11:40:08.051971500 +0800
@@ -65,6 +65,7 @@
 char *pptp_server = NULL;
 char *pptp_client = NULL;
 char *pptp_phone = NULL;
+char *pptp_conn_ID = NULL;		/* if user set connection ID/name in UI, record it */
 int pptp_window=50;
 int pptp_sock=-1;
 struct in_addr localbind = { INADDR_NONE };
@@ -89,6 +90,8 @@
       "PPTP Phone number" },
     { "pptp_window",o_int, &pptp_window,
       "PPTP window" },
+    { "pptp_conn_ID", o_string, &pptp_conn_ID,
+      "connection ID/name" },
     { NULL }
 };
 
@@ -147,7 +150,10 @@
 
 	src_addr.sa_family=AF_PPPOX;
 	src_addr.sa_protocol=PX_PROTO_PPTP;
-	src_addr.sa_addr.pptp.call_id=0;
+	if(pptp_conn_ID != NULL)
+		src_addr.sa_addr.pptp.call_id=(atoi(pptp_conn_ID));
+	else
+		src_addr.sa_addr.pptp.call_id=0;
 
 	dst_addr.sa_family=AF_PPPOX;
 	dst_addr.sa_protocol=PX_PROTO_PPTP;

