--- a/include/linux/netfilter_ipv4/ipt_REJECT.h
+++ b/include/linux/netfilter_ipv4/ipt_REJECT.h
@@ -10,7 +10,8 @@ enum ipt_reject_with {
 	IPT_ICMP_NET_PROHIBITED,
 	IPT_ICMP_HOST_PROHIBITED,
 	IPT_TCP_RESET,
-	IPT_ICMP_ADMIN_PROHIBITED
+	IPT_ICMP_ADMIN_PROHIBITED,
+	IPT_HTTP_SITE_PROHIBITED
 };
 
 struct ipt_reject_info {
--- a/extensions/libipt_REJECT.c
+++ b/extensions/libipt_REJECT.c
@@ -53,7 +53,9 @@ static const struct reject_names reject_table[] = {
 	{"tcp-reset", "tcp-rst",
 	 IPT_TCP_RESET, "TCP RST packet"},
 	{"icmp-admin-prohibited", "admin-prohib",
-	 IPT_ICMP_ADMIN_PROHIBITED, "ICMP administratively prohibited (*)"}
+	 IPT_ICMP_ADMIN_PROHIBITED, "ICMP administratively prohibited (*)"},
+	{"http-site-block", "site-block", IPT_HTTP_SITE_PROHIBITED, "http site block"}
+	
 };
 
 static void
@@ -62,7 +64,7 @@ print_reject_types(void)
 	unsigned int i;
 
 	printf("Valid reject types:\n");
-
+    
 	for (i = 0; i < ARRAY_SIZE(reject_table); ++i) {
 		printf("    %-25s\t%s\n", reject_table[i].name, reject_table[i].desc);
 		printf("    %-25s\talias\n", reject_table[i].alias);
@@ -156,7 +158,8 @@ static const struct reject_names_xlate reject_table_xlate[] = {
 	{"net-prohibited",	IPT_ICMP_NET_PROHIBITED},
 	{"host-prohibited",	IPT_ICMP_HOST_PROHIBITED},
 	{"tcp reset",		IPT_TCP_RESET},
-	{"admin-prohibited",	IPT_ICMP_ADMIN_PROHIBITED}
+	{"admin-prohibited",	IPT_ICMP_ADMIN_PROHIBITED},
+	{"site-block",          IPT_HTTP_SITE_PROHIBITED}
 };
 
 static int REJECT_xlate(struct xt_xlate *xl,

