--- a/switch_fun.c
+++ b/switch_fun.c
@@ -2135,7 +2135,7 @@ void igmp_on(int argc, char *argv[])
             /* reg_write(0x2010, 0x810000c8); */
             reg_read(0x2510, &value);
             reg_write(0x2510, value | (1 << 3));
-            reg_write(REG_ISC_ADDR, 0x01027d01);
+            reg_write(REG_ISC_ADDR, 0x20027d60);
         }
     }
     else
@@ -2150,120 +2150,128 @@ void igmp_on(int argc, char *argv[])
 	reg_write(0x2308, 0xb3ff);
 	reg_write(0x2408, 0xb3ff);
 	reg_write(0x2608, 0xb3ff);
-	/* Enable Port ACL
-	* reg_write(0x2P04, 0xff0403);
-	*/
-	for (port = 0; port <= 6; port++) {
-		offset = 0x2004 + port * 0x100;
-		reg_read(offset, &value);
-		reg_write(offset, value | (1 << 10));
-	}
-
-	/*IGMP query only p4 -> p5*/
-	reg_write(0x94, 0x00ff0002);
-    if (wan_num == 4)
-    {   
-        reg_write(0x98, 0x000a1008);
-    }
-    else if (wan_num == 0)
-    {   
-        reg_write(0x98, 0x000a0108);
-    }
-    else if (wan_num == 5)
-    {
-        reg_write(0x98, 0x000a2008);
-    }
-	reg_write(0x90, 0x80005000);
-	reg_write(0x94, 0xff001100);
-    if (wan_num == 4)
-    {   
-        reg_write(0x98, 0x000B1000);
-    }
-    else if (wan_num == 0)
-    {   
-        reg_write(0x98, 0x000B0100);
-    }
-    else if (wan_num == 5)
-    {
-        reg_write(0x98, 0x000B2000);
-    }
-    
-	reg_write(0x90, 0x80005001);
-	reg_write(0x94, 0x3);
-	reg_write(0x98, 0x0);
-	reg_write(0x90, 0x80009000);
-	reg_write(0x94, 0x1a002080);
-	reg_write(0x98, 0x0);
-	reg_write(0x90, 0x8000b000);
-
-	/*IGMP p5 -> p4*/
-	reg_write(0x94, 0x00ff0002);
-	reg_write(0x98, 0x000a2008);
-	reg_write(0x90, 0x80005002);
-	reg_write(0x94, 0x4);
-	reg_write(0x98, 0x0);
-	reg_write(0x90, 0x80009001);
-    if (wan_num == 4)
-    {   
-        reg_write(0x94, 0x1a001080);
-    }
-    else if (wan_num == 0)
-    {   
-        reg_write(0x94, 0x1a000180);
-    }
-    else if (wan_num == 5)
-    {
-        reg_write(0x94, 0x1a002080);
-    }
-	reg_write(0x98, 0x0);
-	reg_write(0x90, 0x8000b001);
 
-	/*IGMP p0~p3 -> p6*/
-	reg_write(0x94, 0x00ff0002);
-    if (wan_num == 4)
-    {   
-        reg_write(0x98, 0x000a0f08);
-    }
+	if (wan_num == 5)
+	{
+		reg_write(0x2508, 0xb3ff);
+	}
 	else if (wan_num == 0)
-    {   
-		reg_write(0x98, 0x000a1e08);
-    }
-    else if (wan_num == 5)
-    {
-       reg_write(0x98, 0x000a1f08);
-    }
-	reg_write(0x90, 0x80005003);
-	reg_write(0x94, 0x8);
-	reg_write(0x98, 0x0);
-	reg_write(0x90, 0x80009002);
-	reg_write(0x94, 0x1a004080);
-	reg_write(0x98, 0x0);
-	reg_write(0x90, 0x8000b002);
-
-	/*IGMP query only p6 -> p0~p3*/
-	reg_write(0x94, 0x00ff0002);
-	reg_write(0x98, 0x000a4008);
-	reg_write(0x90, 0x80005004);
-	reg_write(0x94, 0xff001100);
-	reg_write(0x98, 0x000B4000);
-	reg_write(0x90, 0x80005005);
-	reg_write(0x94, 0x30);
-	reg_write(0x98, 0x0);
-	reg_write(0x90, 0x80009003);
-    if (wan_num == 4)
-    {
-        reg_write(0x94, 0x1a000f80);
-    }
-    else if (wan_num == 0)
-    {
-        reg_write(0x94, 0x1a001e80);
-    }
-    else if (wan_num == 5)
-    {
-        reg_write(0x94, 0x1a001f80);
-    }
-	reg_write(0x98, 0x0);
-	reg_write(0x90, 0x8000b003);
+	{
+		/* Enable Port ACL
+		* reg_write(0x2P04, 0xff0403);
+		*/
+		for (port = 0; port <= 6; port++) {
+			offset = 0x2004 + port * 0x100;
+			reg_read(offset, &value);
+			reg_write(offset, value | (1 << 10));
+		}
+
+		/*IGMP query only p4 -> p5*/
+		reg_write(0x94, 0x00ff0002);
+		if (wan_num == 4)
+		{   
+			reg_write(0x98, 0x000a1008);
+		}
+		else if (wan_num == 0)
+		{   
+			reg_write(0x98, 0x000a0108);
+		}
+		else if (wan_num == 5)
+		{
+			reg_write(0x98, 0x000a2008);
+		}
+		reg_write(0x90, 0x80005000);
+		reg_write(0x94, 0xff001100);
+		if (wan_num == 4)
+		{   
+			reg_write(0x98, 0x000B1000);
+		}
+		else if (wan_num == 0)
+		{   
+			reg_write(0x98, 0x000B0100);
+		}
+		else if (wan_num == 5)
+		{
+			reg_write(0x98, 0x000B2000);
+		}
+		
+		reg_write(0x90, 0x80005001);
+		reg_write(0x94, 0x3);
+		reg_write(0x98, 0x0);
+		reg_write(0x90, 0x80009000);
+		reg_write(0x94, 0x1a002080);
+		reg_write(0x98, 0x0);
+		reg_write(0x90, 0x8000b000);
+
+		/*IGMP p5 -> p4*/
+		reg_write(0x94, 0x00ff0002);
+		reg_write(0x98, 0x000a2008);
+		reg_write(0x90, 0x80005002);
+		reg_write(0x94, 0x4);
+		reg_write(0x98, 0x0);
+		reg_write(0x90, 0x80009001);
+		if (wan_num == 4)
+		{   
+			reg_write(0x94, 0x1a001080);
+		}
+		else if (wan_num == 0)
+		{   
+			reg_write(0x94, 0x1a000180);
+		}
+		else if (wan_num == 5)
+		{
+			reg_write(0x94, 0x1a002080);
+		}
+		reg_write(0x98, 0x0);
+		reg_write(0x90, 0x8000b001);
+
+		/*IGMP p0~p3 -> p6*/
+		reg_write(0x94, 0x00ff0002);
+		if (wan_num == 4)
+		{   
+			reg_write(0x98, 0x000a0f08);
+		}
+		else if (wan_num == 0)
+		{   
+			reg_write(0x98, 0x000a1e08);
+		}
+		else if (wan_num == 5)
+		{
+		reg_write(0x98, 0x000a1f08);
+		}
+		reg_write(0x90, 0x80005003);
+		reg_write(0x94, 0x8);
+		reg_write(0x98, 0x0);
+		reg_write(0x90, 0x80009002);
+		reg_write(0x94, 0x1a004080);
+		reg_write(0x98, 0x0);
+		reg_write(0x90, 0x8000b002);
+
+		/*IGMP query only p6 -> p0~p3*/
+		reg_write(0x94, 0x00ff0002);
+		reg_write(0x98, 0x000a4008);
+		reg_write(0x90, 0x80005004);
+		reg_write(0x94, 0xff001100);
+		reg_write(0x98, 0x000B4000);
+		reg_write(0x90, 0x80005005);
+		reg_write(0x94, 0x30);
+		reg_write(0x98, 0x0);
+		reg_write(0x90, 0x80009003);
+		if (wan_num == 4)
+		{
+			reg_write(0x94, 0x1a000f80);
+		}
+		else if (wan_num == 0)
+		{
+			reg_write(0x94, 0x1a001e80);
+		}
+		else if (wan_num == 5)
+		{
+			reg_write(0x94, 0x1a001f80);
+		}
+		reg_write(0x98, 0x0);
+		reg_write(0x90, 0x8000b003);
+	}
 
 	/*Force eth2 to receive all igmp packets*/
 	sprintf(cmd,"echo 2 > /sys/devices/virtual/net/%s/brif/%s.0/multicast_router",BR_DEVNAME,ETH_DEVNAME);
