Index: linux-4.4.198/drivers/net/phy/mtk/mt753x/mt753x.h
===================================================================
--- linux-4.4.198.orig/drivers/net/phy/mtk/mt753x/mt753x.h
+++ linux-4.4.198/drivers/net/phy/mtk/mt753x/mt753x.h
@@ -24,7 +24,7 @@
 #define MT753X_DFL_CPU_PORT	6
 #define MT753X_NUM_PORTS	7
 #define MT753X_NUM_PHYS		5
-#define MT753X_NUM_VLANS	4095
+#define MT753X_NUM_VLANS	4096
 
 #define MT753X_MAX_VID		4095
 #define MT753X_MIN_VID		0
