--- a/drivers/leds/leds-gpio.c	2020-02-03 09:08:23.000000000 +0800
+++ b/drivers/leds/leds-gpio.c	2020-09-30 16:46:57.202662414 +0800
@@ -22,9 +22,15 @@
 #include <linux/slab.h>
 #include <linux/workqueue.h>
 
+#define COMBINED_LED_SIZE 2
 struct gpio_led_data {
 	struct led_classdev cdev;
-	struct gpio_desc *gpiod;
+	u32 bcombined;
+	u32 combined_num;
+	union {
+		struct gpio_desc *gpiod;
+		struct gpio_led_data *gpio_combined[COMBINED_LED_SIZE];
+	};
 	struct work_struct work;
 	u8 new_level;
 	u8 can_sleep;
@@ -38,11 +44,31 @@
 		container_of(work, struct gpio_led_data, work);
 
 	if (led_dat->blinking) {
-		led_dat->platform_gpio_blink_set(led_dat->gpiod,
+		if (led_dat->bcombined) {
+			int idx = 0;
+			while (idx < led_dat->combined_num) {
+				struct gpio_led_data *tmp_led_dat = led_dat->gpio_combined[idx];
+				tmp_led_dat->platform_gpio_blink_set(tmp_led_dat->gpiod, led_dat->new_level,
+						NULL, NULL);
+				idx++;
+			}
+		} else {
+			led_dat->platform_gpio_blink_set(led_dat->gpiod,
 					led_dat->new_level, NULL, NULL);
+		}
 		led_dat->blinking = 0;
-	} else
-		gpiod_set_value_cansleep(led_dat->gpiod, led_dat->new_level);
+	} else {
+		if (led_dat->bcombined) {
+			int idx = 0;
+			while (idx < led_dat->combined_num) {
+				struct gpio_led_data *tmp_led_dat = led_dat->gpio_combined[idx];
+				gpiod_set_value_cansleep(tmp_led_dat->gpiod, led_dat->new_level);
+				idx++;
+			}
+		} else {
+			gpiod_set_value_cansleep(led_dat->gpiod, led_dat->new_level);
+		}
+	}
 }
 
 static void gpio_led_set(struct led_classdev *led_cdev,
@@ -66,11 +92,31 @@
 		schedule_work(&led_dat->work);
 	} else {
 		if (led_dat->blinking) {
+			if (led_dat->bcombined) {
+				int idx = 0;
+				while (idx < led_dat->combined_num) {
+					struct gpio_led_data *tmp_led_dat = led_dat->gpio_combined[idx];
+					tmp_led_dat->platform_gpio_blink_set(tmp_led_dat->gpiod, level,
+                                                         NULL, NULL);
+					idx++;
+				}
+			} else {
 			led_dat->platform_gpio_blink_set(led_dat->gpiod, level,
 							 NULL, NULL);
+			}
 			led_dat->blinking = 0;
-		} else
-			gpiod_set_value(led_dat->gpiod, level);
+		} else {
+			if (led_dat->bcombined) {
+                                int idx = 0;
+                                while (idx < led_dat->combined_num) {
+                                        struct gpio_led_data *tmp_led_dat = led_dat->gpio_combined[idx];
+                                        gpiod_set_value(tmp_led_dat->gpiod, level);
+                                        idx++;
+                                }
+                        } else {
+				gpiod_set_value(led_dat->gpiod, level);
+			}
+		}
 	}
 }
 
@@ -81,8 +127,19 @@
 		container_of(led_cdev, struct gpio_led_data, cdev);
 
 	led_dat->blinking = 1;
-	return led_dat->platform_gpio_blink_set(led_dat->gpiod, GPIO_LED_BLINK,
-						delay_on, delay_off);
+	if (led_dat->bcombined) {
+		int idx = 0;
+		while (idx < led_dat->combined_num) {
+			struct gpio_led_data *tmp_led_dat = led_dat->gpio_combined[idx];
+			tmp_led_dat->platform_gpio_blink_set(tmp_led_dat->gpiod, GPIO_LED_BLINK,
+					delay_on, delay_off);
+			idx++;
+		}
+		return 0;
+	} else {
+		return led_dat->platform_gpio_blink_set(led_dat->gpiod, GPIO_LED_BLINK,
+				delay_on, delay_off);
+	}
 }
 
 static int create_gpio_led(const struct gpio_led *template,
@@ -91,8 +148,11 @@
 {
 	int ret, state;
 
-	led_dat->gpiod = template->gpiod;
-	if (!led_dat->gpiod) {
+	if(!led_dat->bcombined) {
+		led_dat->gpiod = template->gpiod;
+	}
+
+	if (!led_dat->bcombined && !led_dat->gpiod) {
 		/*
 		 * This is the legacy code path for platform code that
 		 * still uses GPIO numbers. Ultimately we would like to get
@@ -122,7 +182,15 @@
 
 	led_dat->cdev.name = template->name;
 	led_dat->cdev.default_trigger = template->default_trigger;
-	led_dat->can_sleep = gpiod_cansleep(led_dat->gpiod);
+	if (led_dat->bcombined) {
+		int idx = 0;
+		while (idx < led_dat->combined_num) {
+			led_dat->can_sleep &= gpiod_cansleep(led_dat->gpio_combined[idx]->gpiod);
+			idx++;
+		}
+	} else {
+		led_dat->can_sleep = gpiod_cansleep(led_dat->gpiod);
+	}
 	led_dat->blinking = 0;
 	if (blink_set) {
 		led_dat->platform_gpio_blink_set = blink_set;
@@ -138,10 +206,12 @@
 		led_dat->cdev.flags |= LED_CORE_SUSPENDRESUME;
 	if (template->panic_indicator)
 		led_dat->cdev.flags |= LED_PANIC_INDICATOR;
-
+	
+	if(!led_dat->bcombined) {
 	ret = gpiod_direction_output(led_dat->gpiod, state);
 	if (ret < 0)
 		return ret;
+	}
 
 	INIT_WORK(&led_dat->work, gpio_led_work);
 
@@ -184,12 +254,35 @@
 	device_for_each_child_node(dev, child) {
 		struct gpio_led led = {};
 		const char *state = NULL;
+		const char *combined_names[COMBINED_LED_SIZE];
+		int combined_num = 0;
 
-		led.gpiod = devm_get_gpiod_from_child(dev, NULL, child);
-		if (IS_ERR(led.gpiod)) {
-			fwnode_handle_put(child);
-			ret = PTR_ERR(led.gpiod);
-			goto err;
+		if (fwnode_property_present(child, "combined")) {
+			int name_idx, led_idx;
+			priv->leds[priv->num_leds].bcombined = true;
+			combined_num = fwnode_property_read_string_array(child, "combined", combined_names, COMBINED_LED_SIZE);
+			name_idx = 0;
+			while (name_idx < combined_num) {
+				led_idx = 0;
+				while (led_idx < priv->num_leds) {
+					struct gpio_led_data *led_data = &priv->leds[priv->num_leds];
+					struct led_classdev *cdev = &priv->leds[led_idx].cdev;
+					if (!strcmp(combined_names[name_idx], cdev->name)) {
+						led_data->gpio_combined[led_data->combined_num] = &priv->leds[led_idx];
+						led_data->combined_num ++;
+					}
+					led_idx++;
+				}
+				name_idx++;
+			}
+		} else {
+			priv->leds[priv->num_leds].bcombined = false;
+			led.gpiod = devm_get_gpiod_from_child(dev, NULL, child);
+			if (IS_ERR(led.gpiod)) {
+				fwnode_handle_put(child);
+				ret = PTR_ERR(led.gpiod);
+				goto err;
+			}
 		}
 
 		np = to_of_node(child);
