--- a/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c
+++ b/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c
@@ -232,8 +232,21 @@ int nf_hnat_netdevice_event(struct notif
 		foe_clear_all_bind_entries(dev);
 
 		break;
-	default:
+	case NETDEV_UNREGISTER:
+		if (IS_PPD(dev) && hnat_priv->g_ppdev)
+		{
+			hnat_priv->g_ppdev = NULL;
+			dev_put(dev);
+		}
+		break;
+	case NETDEV_REGISTER:
+		if (IS_PPD(dev) && !hnat_priv->g_ppdev)
+		{
+			hnat_priv->g_ppdev = dev_get_by_name(&init_net, hnat_priv->ppd);
+		}
 		break;
+	default:
+	break;
 	}
 
 	return NOTIFY_DONE;
